urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/framework/core/Exec.java

```diff
@@ -0,0 +1,259 @@
+                ok = true;
+                if(m_verbose) System.out.println("Controller initialization time: " + (PTSPConstants.getStepsInit(m_game.getNumWaypoints()) + timeAfter - finalDateMs) + " ms.");
+            }
+        }catch(InstantiationException e)
+        {
+            System.err.println("Exception instantiating " + a_contName + ":");
+            e.printStackTrace();
+        }catch(IllegalAccessException e)
+        {
+            System.err.println("Illegal access exception when instantiating " + a_contName + ":");
+            e.printStackTrace();
+        }catch(InvocationTargetException e)
+        {
+            System.err.println("Exception calling the constructor " + a_contName + "(Game):");
+            e.printStackTrace();
+        }
+
+        return ok;
+    }
+
+    /**
+     * Creates a game and the controller.
+     * @return true if the game is ready to be played.
+     */
+    protected static boolean prepareGame()
+    {
+        //Create the game instance.
+        m_game = new Game(m_mapNames);
+        //and the controller
+        return instanceController();
+    }
+
+    /**
+     * Waits until the next step.
+     * @param duration Amount of time to wait for.
+     */
+    protected static void waitStep(int duration) {
+
+        try
+        {
+            Thread.sleep(duration);
+        }
+        catch(InterruptedException e)
+        {
+            e.printStackTrace();
+        }
+    }
+
+    protected static boolean runFromData( char[][] map, Vector2d startingPoint, LinkedList<Vector2d> wayPoints, LinkedList<Vector2d> fuelTanks)
+    {
+        //Create the game instance.
+        m_game = new Game(map, startingPoint, wayPoints, fuelTanks);
+
+        //and the controller
+        return instanceController();
+    }
+
+
+
+    /**
+     * Writes a map to the given filename.
+     */
+    protected static void writeMap(Map map, String filename)
+    {
+        StringBuffer sb = new StringBuffer();
+        char[][] grid = map.getMapChar();
+        String[] lines = new String[grid[0].length+4];
+
+        lines[0] = "type octile";
+        lines[1] = "height " + grid[0].length;
+        lines[2] = "width " + grid.length;
+        lines[3] = "map";
+
+        for(int i = 0; i < grid[0].length; i++)
+        {
+            for(int j = 0; j < grid.length; ++j)
+            {
+                char data = grid[j][i];
+                sb.append(data);
+            }
+            lines[i+4] = sb.toString();
+            sb = new StringBuffer();
+        }
+
+        File2String.put(lines, filename);
+    }
+
+    /**
+     * Reads the forces from a file and returns them in an array.
+     * @param a_filename Name of the file where the forces are.
+     * @return the array of forces to apply.
+     * @throws Exception from file handling.
+     */
+    protected static int[] readForces(String a_filename) throws Exception {
+        BufferedReader in = new BufferedReader(new FileReader(a_filename));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at lines 246, 247, 251. The resource is closed at line 254. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).